File: scripts/gencode.scala
// Extract code snippets from the complete Chisel examples in src

import java.io._
import scala.io.Source

  list("src/main/scala/")
  list("src/test/scala/")
  list("src/main/vhdl/")

def list(folder: String): Unit = {
  new File(folder).listFiles(_.isFile).map(f => extract(folder + f.getName))
  new File(folder).listFiles(_.isDirectory).map(f => list(folder + f.getName + "/"))
}

def extract(f: String) {
  println(f)
  var code: PrintWriter = null
  val lines = Source.fromFile(f).getLines()
  for (l <- lines) {
    val tokens = l.trim.split(" ")
    if (tokens.length >= 2 && (tokens(0).equals("//-") || tokens(0).equals("--/"))) {
      if (tokens(1).equals("start")) {
        code = new PrintWriter(new File("code/"+tokens(2)+".txt"))
      } else if (tokens(1).equals("end")) {
        code.close()
        code = null
      }
    } else if (code != null) {
        code.println(l)
    }
  }
}

File: scripts/gencode.py
#!/usr/bin/env python3
# Extract code snippets from the complete Chisel examples in src

import os

def list_files(folder):
    for entry in os.listdir(folder):
        full_path = os.path.join(folder, entry)
        if os.path.isfile(full_path):
            extract(full_path)
        elif os.path.isdir(full_path):
            list_files(full_path)

def extract(f):
    print(f)
    code = None
    with open(f, 'r', errors='replace') as infile:
        for line in infile:
            tokens = line.strip().split()
            if len(tokens) >= 2 and tokens[0] in ('//-', '--/'):
                if tokens[1] == 'start':
                    code = open(os.path.join('code', tokens[2] + '.txt'), 'w')
                elif tokens[1] == 'end':
                    if code is not None:
                        code.close()
                        code = None
            elif code is not None:
                code.write(line)
    if code is not None:
        code.close()

list_files('src/main/scala/')
list_files('src/test/scala/')
list_files('src/main/vhdl/')
